src/swap/comit.rs

```diff
@@ -0,0 +1,346 @@
+//! Code that could be upstreamed to COMIT lib.
+
+pub mod hbit;
+pub mod herc20;
+
+use chrono::NaiveDateTime;
+use comit::btsieve::{BlockByHash, LatestBlock};
+use futures::{
+    future::{self, Either},
+    pin_mut,
+};
+
+pub use comit::{ethereum, *};
+
+/// Execute Alice's side of a Hbit<->Herc20 swap.
+#[cfg(test)]
+pub async fn hbit_herc20_alice<A, BC, EC>(
+    alice: A,
+    bitcoin_connector: &BC,
+    ethereum_connector: &EC,
+    hbit_params: hbit::Params,
+    herc20_params: herc20::Params,
+    secret: comit::Secret,
+    start_of_swap: NaiveDateTime,
+) -> anyhow::Result<()>
+where
+    A: hbit::ExecuteFund + hbit::ExecuteRefund + herc20::ExecuteRedeem,
+    BC: LatestBlock<Block = ::bitcoin::Block>
+        + BlockByHash<Block = ::bitcoin::Block, BlockHash = ::bitcoin::BlockHash>,
+    EC: LatestBlock<Block = ethereum::Block>
+        + BlockByHash<Block = ethereum::Block, BlockHash = ethereum::Hash>
+        + btsieve::ethereum::ReceiptByHash,
+{
+    let hbit_funded = match alice.execute_fund(&hbit_params).await {
+        Ok(hbit_funded) => hbit_funded,
+        Err(_) => return Ok(()),
+    };
+
+    let herc20_deployed =
+        match herc20::watch_for_deployed(ethereum_connector, herc20_params.clone(), start_of_swap)
+            .await
+        {
+            Ok(herc20_deployed) => herc20_deployed,
+            Err(_) => {
+                alice.execute_refund(hbit_params, hbit_funded).await?;
+
+                return Ok(());
+            }
+        };
+
+    let _herc20_funded = match herc20::watch_for_funded(
+        ethereum_connector,
+        herc20_params.clone(),
+        start_of_swap,
+        herc20_deployed.clone(),
+    )
+    .await
+    {
+        Ok(herc20_funded) => herc20_funded,
+        Err(_) => {
+            alice.execute_refund(hbit_params, hbit_funded).await?;
+
+            return Ok(());
+        }
+    };
+
+    let _herc20_redeemed = match alice
+        .execute_redeem(herc20_params, secret, herc20_deployed, start_of_swap)
+        .await
+    {
+        Ok(herc20_redeemed) => herc20_redeemed,
+        Err(_) => {
+            alice.execute_refund(hbit_params, hbit_funded).await?;
+
+            return Ok(());
+        }
+    };
+
+    let hbit_redeem = hbit::watch_for_redeemed(
+        bitcoin_connector,
+        &hbit_params.shared,
+        hbit_funded.location,
+        start_of_swap,
+    );
+    let hbit_refund = alice.execute_refund(hbit_params, hbit_funded);
```

Comment:
Alice already redeemed at this stage no (line 68)? Not really a point in executing refund :)

Reply:
Given that this is in the body of `hbit_herc20_alice`, I think she should always try to refund after she's done redeeming the ERC20 tokens. If anything to make sure that someone claims the money in case Bob fails to.

On the other hand, I think we should actually not care whether Bob redeems or not. From Alice's perspective at this stage she only cares about maybe being able to refund for free.

In conclusion, for swap executions involving Alice, I'm just going to call `alice.execute_refund` at the end.

Reply:
> In conclusion, for swap executions involving Alice, I'm just going to call alice.execute_refund at the end.

In practice this doesn't work so well, because we may end up waiting forever for this possible refund. In Nectar we rely on knowing when the swap execution ends to free funds and save the swap as a trade to a history file.

This could be solved by simply spawning the refund task away, but this would add an implicit (or explicit, if we decided to pass a handle as an argument) dependency on `tokio::Runtime`.

Thoughts, @thomaseizinger?

Reply:
Though questions :D

It really boils down to application policy I guess? Are you gonna try and claim money that actually belongs to the other party? I am tempted to say we should just wait for the other party to redeem. In the end, the coins will either go to Alice or Bob, none else can take them so there isn't really a risk of "OMG free money someone's gotta take it :D"

> This could be solved by simply spawning the refund task away, but this would add an implicit (or explicit, if we decided to pass a handle as an argument) dependency on tokio::Runtime.

Alternatively, you could return the future from this function :)

Reply:
> I am tempted to say we should just wait for the other party to redeem. In the end, the coins will either go to Alice or Bob, none else can take them so there isn't really a risk of "OMG free money someone's gotta take it :D"

Yeah, I guess the library shouldn't enforce taking the other party's money if they mess up. Instead of waiting for Bob to redeem though, I would just consider the protocol done from Alice's perspective and exit.